File: backend/app/routers/stock_token.py
from __future__ import annotations

from datetime import timedelta
from decimal import Decimal, InvalidOperation, ROUND_DOWN
from math import ceil
from typing import Any, Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from app.db.models.stock_token_convert_record import StockTokenConvertRecord
from app.db.models.stock_token_lock_config import StockTokenLockConfig
from app.db.models.user_stock_token_lock import UserStockTokenLock
from app.db.session import get_db
from app.deps.auth import get_current_user_id
from app.services.stock_token_lock_service import (
    StockTokenLockError,
    calculate_stock_token_release_progress,
    convert_released_stock_token,
    get_stock_token_unlock_at,
    is_stock_token_lock_in_lock_period,
)


router = APIRouter(prefix="/stock-token", tags=["stock-token"])
Q18 = Decimal("0.000000000000000001")


class StockTokenConvertIn(BaseModel):
    lock_item_id: int = Field(..., ge=1)
    amount: str


def _decimal_to_str(value: Any) -> str:
    amount = value if isinstance(value, Decimal) else Decimal(str(value or "0"))
    return format(amount.quantize(Q18, rounding=ROUND_DOWN), "f")


def _datetime_to_str(value: Any) -> Optional[str]:
    if value is None:
        return None
    if hasattr(value, "isoformat"):
        return value.isoformat()
    return str(value)


def _stock_token_release_days(daily_release_rate: Any) -> int:
    try:
        rate = Decimal(str(daily_release_rate or "0"))
    except (InvalidOperation, TypeError, ValueError):
        return 0
    if rate <= Decimal("0"):
        return 0
    return max(1, int(ceil(Decimal("1") / rate)))


def _parse_decimal_amount(value: str) -> Decimal:
    try:
        return Decimal(str(value).strip())
    except (InvalidOperation, TypeError, ValueError):
        raise HTTPException(
            status_code=400,
            detail={"code": "INVALID_AMOUNT", "message": "amount 格式不正确"},
        )


@router.get("/locks")
def get_my_stock_token_locks(
    user_id: int = Depends(get_current_user_id),
    db: Session = Depends(get_db),
):
    rows = (
        db.query(UserStockTokenLock, StockTokenLockConfig)
        .outerjoin(StockTokenLockConfig, StockTokenLockConfig.id == UserStockTokenLock.config_id)
        .filter(UserStockTokenLock.user_id == int(user_id))
        .order_by(UserStockTokenLock.created_at.desc(), UserStockTokenLock.id.desc())
        .all()
    )
    items = []
    for lock_item, config in rows:
        in_lock_period = is_stock_token_lock_in_lock_period(lock_item, config=config)
        lock_days = int(config.lock_days or 0) if config else 0
        daily_release_rate = getattr(lock_item, "daily_release_rate_snapshot", None) or (
            config.daily_release_rate if config else 0
        )
        release_days = _stock_token_release_days(daily_release_rate)
        lock_start_at = lock_item.start_at
        lock_end_at = (lock_start_at + timedelta(days=lock_days)) if lock_start_at and lock_days > 0 else None
        release_start_at = lock_end_at
        release_finish_at = (
            release_start_at + timedelta(days=release_days)
            if release_start_at and release_days > 0
            else None
        )
        items.append(
            {
                "id": int(lock_item.id),
                "lock_symbol": lock_item.lock_symbol,
                "trade_symbol": config.trade_symbol if config else None,
                "total_amount": _decimal_to_str(lock_item.total_amount),
                "locked_amount": _decimal_to_str(lock_item.locked_amount),
                "available_amount": _decimal_to_str(lock_item.available_amount),
                "converted_amount": _decimal_to_str(lock_item.converted_amount),
                "conversion_rate_snapshot": _decimal_to_str(
                    getattr(lock_item, "conversion_rate_snapshot", None) or (config.conversion_rate if config else 1)
                ),
                "daily_release_rate": _decimal_to_str(daily_release_rate),
                "lock_days": lock_days,
                "release_days": release_days,
                "unlock_at": _datetime_to_str(get_stock_token_unlock_at(lock_item, config)),
                "lock_start_at": _datetime_to_str(lock_start_at),
                "lock_end_at": _datetime_to_str(lock_end_at),
                "release_start_at": _datetime_to_str(release_start_at),
                "release_finish_at": _datetime_to_str(release_finish_at),
                "release_started": not in_lock_period,
                "progress_percent": str(calculate_stock_token_release_progress(lock_item, config=config)),
                "status": "LOCKED" if in_lock_period and str(lock_item.status or "").upper() == "ACTIVE" else lock_item.status,
                "start_at": _datetime_to_str(lock_item.start_at),
                "end_at": _datetime_to_str(lock_item.end_at),
            }
        )
    return {"items": items}


@router.get("/converts")
def get_my_stock_token_converts(
    user_id: int = Depends(get_current_user_id),
    db: Session = Depends(get_db),
):
    rows = (
        db.query(StockTokenConvertRecord)
        .filter(StockTokenConvertRecord.user_id == int(user_id))
        .order_by(StockTokenConvertRecord.created_at.desc(), StockTokenConvertRecord.id.desc())
        .limit(200)
        .all()
    )
    return {
        "items": [
            {
                "id": int(record.id),
                "from_symbol": record.from_symbol,
                "to_symbol": record.to_symbol,
                "from_amount": _decimal_to_str(record.from_amount),
                "to_amount": _decimal_to_str(record.to_amount),
                "conversion_rate": _decimal_to_str(record.conversion_rate),
                "status": record.status,
                "created_at": _datetime_to_str(record.created_at),
            }
            for record in rows
        ]
    }


@router.post("/convert")
def convert_stock_token(
    payload: StockTokenConvertIn,
    user_id: int = Depends(get_current_user_id),
    db: Session = Depends(get_db),
):
    amount = _parse_decimal_amount(payload.amount)
    try:
        record = convert_released_stock_token(
            db,
            user_id=int(user_id),
            lock_item_id=payload.lock_item_id,
            amount=amount,
        )
        db.commit()
        db.refresh(record)
        return {
            "success": True,
            "record_id": int(record.id),
            "from_symbol": record.from_symbol,
            "to_symbol": record.to_symbol,
            "from_amount": _decimal_to_str(record.from_amount),
            "to_amount": _decimal_to_str(record.to_amount),
        }
    except StockTokenLockError as exc:
        db.rollback()
        raise HTTPException(status_code=400, detail={"code": "STOCK_TOKEN_CONVERT_ERROR", "message": str(exc)})
    except HTTPException:
        db.rollback()
        raise
    except Exception:
        db.rollback()
        raise HTTPException(status_code=500, detail={"code": "INTERNAL_ERROR", "message": "stock token convert failed"})

File: mobile/src/api/stockToken.ts
import {apiClient, type ApiRequestOptions} from './client';
import {normalizeNonNegativeDecimalText} from '../utils/decimalText';

const LOCK_STATUSES = new Set(['LOCKED', 'ACTIVE', 'RELEASED']);
const CONVERT_STATUSES = new Set(['SUCCESS', 'FAILED']);

export type StockTokenLockStatus = 'LOCKED' | 'ACTIVE' | 'RELEASED';
export type StockTokenConvertStatus = 'SUCCESS' | 'FAILED';

export type StockTokenLock = {
  id: number;
  lockSymbol: string;
  tradeSymbol: string | null;
  totalAmount: string;
  lockedAmount: string;
  availableAmount: string;
  convertedAmount: string;
  conversionRateSnapshot: string;
  dailyReleaseRate: string;
  lockDays: number;
  releaseDays: number;
  unlockAt: string | null;
  lockStartAt: string | null;
  lockEndAt: string | null;
  releaseStartAt: string | null;
  releaseFinishAt: string | null;
  releaseStarted: boolean;
  progressPercent: string;
  status: StockTokenLockStatus;
  startAt: string | null;
  endAt: string | null;
};

export type StockTokenConvertRecord = {
  id: number;
  fromSymbol: string;
  toSymbol: string;
  fromAmount: string;
  toAmount: string;
  conversionRate: string;
  status: StockTokenConvertStatus;
  createdAt: string | null;
};

export type StockTokenLocksResponse = {items: StockTokenLock[]};
export type StockTokenConvertsResponse = {items: StockTokenConvertRecord[]};

export class StockTokenContractError extends Error {
  constructor() {
    super('股票代币数据格式异常，请稍后重试');
    this.name = 'StockTokenContractError';
  }
}

export async function fetchStockTokenLocks(
  options?: ApiRequestOptions,
): Promise<StockTokenLocksResponse> {
  const payload = await apiClient.get<unknown>('/stock-token/locks', options);
  return normalizeStockTokenLocks(payload);
}

export async function fetchStockTokenConverts(
  options?: ApiRequestOptions,
): Promise<StockTokenConvertsResponse> {
  const payload = await apiClient.get<unknown>('/stock-token/converts', options);
  return normalizeStockTokenConverts(payload);
}

export function normalizeStockTokenLocks(
  payload: unknown,
): StockTokenLocksResponse {
  const data = requiredRecord(payload);
  if (!Array.isArray(data.items)) throw new StockTokenContractError();
  return {items: data.items.map(normalizeStockTokenLock)};
}

export function normalizeStockTokenConverts(
  payload: unknown,
): StockTokenConvertsResponse {
  const data = requiredRecord(payload);
  if (!Array.isArray(data.items)) throw new StockTokenContractError();
  return {items: data.items.map(normalizeStockTokenConvert)};
}

function normalizeStockTokenLock(payload: unknown): StockTokenLock {
  const data = requiredRecord(payload);
  return {
    id: requiredPositiveInteger(data.id),
    lockSymbol: requiredSymbol(data.lock_symbol),
    tradeSymbol: nullableSymbol(data.trade_symbol),
    totalAmount: requiredDecimal(data.total_amount),
    lockedAmount: requiredDecimal(data.locked_amount),
    availableAmount: requiredDecimal(data.available_amount),
    convertedAmount: requiredDecimal(data.converted_amount),
    conversionRateSnapshot: requiredDecimal(data.conversion_rate_snapshot),
    dailyReleaseRate: requiredDecimal(data.daily_release_rate),
    lockDays: requiredNonNegativeInteger(data.lock_days),
    releaseDays: requiredNonNegativeInteger(data.release_days),
    unlockAt: nullableString(data.unlock_at),
    lockStartAt: nullableString(data.lock_start_at),
    lockEndAt: nullableString(data.lock_end_at),
    releaseStartAt: nullableString(data.release_start_at),
    releaseFinishAt: nullableString(data.release_finish_at),
    releaseStarted: requiredBoolean(data.release_started),
    progressPercent: requiredDecimal(data.progress_percent),
    status: requiredStatus(data.status, LOCK_STATUSES) as StockTokenLockStatus,
    startAt: nullableString(data.start_at),
    endAt: nullableString(data.end_at),
  };
}

function normalizeStockTokenConvert(
  payload: unknown,
): StockTokenConvertRecord {
  const data = requiredRecord(payload);
  return {
    id: requiredPositiveInteger(data.id),
    fromSymbol: requiredSymbol(data.from_symbol),
    toSymbol: requiredSymbol(data.to_symbol),
    fromAmount: requiredDecimal(data.from_amount),
    toAmount: requiredDecimal(data.to_amount),
    conversionRate: requiredDecimal(data.conversion_rate),
    status: requiredStatus(
      data.status,
      CONVERT_STATUSES,
    ) as StockTokenConvertStatus,
    createdAt: nullableString(data.created_at),
  };
}

function requiredRecord(value: unknown): Record<string, unknown> {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new StockTokenContractError();
  }
  return value as Record<string, unknown>;
}

function requiredString(value: unknown) {
  if (typeof value !== 'string' || !value.trim()) {
    throw new StockTokenContractError();
  }
  return value.trim();
}

function nullableString(value: unknown) {
  if (value === null || value === undefined) return null;
  return requiredString(value);
}

function requiredSymbol(value: unknown) {
  const symbol = requiredString(value).toUpperCase();
  if (!/^[A-Z0-9._-]{1,30}$/.test(symbol)) {
    throw new StockTokenContractError();
  }
  return symbol;
}

function nullableSymbol(value: unknown) {
  if (value === null || value === undefined) return null;
  return requiredSymbol(value);
}

function requiredDecimal(value: unknown) {
  const normalized = normalizeNonNegativeDecimalText(value);
  if (normalized === null) throw new StockTokenContractError();
  return normalized;
}

function requiredBoolean(value: unknown) {
  if (typeof value !== 'boolean') throw new StockTokenContractError();
  return value;
}

function requiredNonNegativeInteger(value: unknown) {
  if (!Number.isSafeInteger(value) || Number(value) < 0) {
    throw new StockTokenContractError();
  }
  return Number(value);
}

function requiredPositiveInteger(value: unknown) {
  const number = requiredNonNegativeInteger(value);
  if (number < 1) throw new StockTokenContractError();
  return number;
}

function requiredStatus(value: unknown, allowed: Set<string>) {
  const status = requiredString(value).toUpperCase();
  if (!allowed.has(status)) throw new StockTokenContractError();
  return status;
}
